allennlp/modules/span_extractors/max_pooling_span_extractor.py

```diff
@@ -0,0 +1,131 @@
+    of the resulting span s is computed via s_d = max(x_id, ..., x_jd).
+
+    Elements masked-out by sequence_mask are ignored when max-pooling is computed.
+    Span representations of masked out span_indices by span_mask are set to '0.'
+
+    Registered as a `SpanExtractor` with name "max_pooling".
+
+    # Parameters
+
+    input_dim : `int`, required.
+        The final dimension of the `sequence_tensor`.
+    num_width_embeddings : `int`, optional (default = `None`).
+        Specifies the number of buckets to use when representing
+        span width features.
+    span_width_embedding_dim : `int`, optional (default = `None`).
+        The embedding size for the span_width features.
+    bucket_widths : `bool`, optional (default = `False`).
+        Whether to bucket the span widths into log-space buckets. If `False`,
+        the raw span widths are used.
+
+    # Returns
+
+    max_pooling_text_embeddings : `torch.FloatTensor`.
+        A tensor of shape (batch_size, num_spans, input_dim), which each span representation
+        is the result of a max-pooling operation.
+
+    """
+
+    def __init__(
+        self,
+        input_dim: int,
+        num_width_embeddings: int = None,
+        span_width_embedding_dim: int = None,
+        bucket_widths: bool = False,
+    ) -> None:
+        super().__init__(
+            input_dim=input_dim,
+            num_width_embeddings=num_width_embeddings,
+            span_width_embedding_dim=span_width_embedding_dim,
+            bucket_widths=bucket_widths,
+        )
+
+    def get_output_dim(self) -> int:
+        if self._span_width_embedding is not None:
+            return self._input_dim + self._span_width_embedding.get_output_dim()
+        return self._input_dim
+
+    def _embed_spans(
+        self,
+        sequence_tensor: torch.FloatTensor,
+        span_indices: torch.LongTensor,
+        sequence_mask: torch.BoolTensor = None,
+        span_indices_mask: torch.BoolTensor = None,
+    ) -> torch.FloatTensor:
+
+        if sequence_tensor.size(-1) != self._input_dim:
+            raise ValueError(
+                f"Dimension mismatch expected ({sequence_tensor.size(-1)}) "
+                f"received ({self._input_dim})."
+            )
+
+        if sequence_tensor.shape[1] <= span_indices.max() or span_indices.min() < 0:
+            raise IndexError(
+                f"Span index out of range, max index ({span_indices.max()}) "
+                f"or min index ({span_indices.min()}) "
+                f"not valid for sequence of length ({sequence_tensor.shape[1]})."
+            )
+
+        if (span_indices[:, :, 0] > span_indices[:, :, 1]).any():
+            raise IndexError(
+                "Span start above span end",
+            )
+
+        # Calculate the maximum sequence length for each element in batch.
+        # If span_end indices are above these length, we adjust the indices in adopted_span_indices
```

Comment:
I think a better name would be "adapted_span_indices" instead of "adopted_..."
```suggestion
        # If span_end indices are above these length, we adjust the indices in adapted_span_indices
```